crates/sui-types/src/crypto.rs

```diff
@@ -33,21 +35,129 @@ use crate::error::{SuiError, SuiResult};
+
+impl signature::Signer<Signature> for SuiKeyPair {
+    fn try_sign(&self, msg: &[u8]) -> Result<Signature, signature::Error> {
+        match self {
+            SuiKeyPair::Ed25519KeyPair(kp) => kp.try_sign(msg),
+            SuiKeyPair::Secp256k1KeyPair(kp) => kp.try_sign(msg),
+        }
+    }
+}
+
+impl FromStr for SuiKeyPair {
+    type Err = eyre::Report;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let kp = Self::decode_base64(s).map_err(|e| eyre::eyre!("{}", e.to_string()))?;
+        Ok(kp)
+    }
+}
+
+impl EncodeDecodeBase64 for SuiKeyPair {
+    fn encode_base64(&self) -> String {
+        let mut bytes: Vec<u8> = Vec::new();
+        match self {
+            SuiKeyPair::Ed25519KeyPair(kp) => {
+                let kp1 = kp.copy();
+                bytes.extend_from_slice(&[self.public().flag()]);
+                bytes.extend_from_slice(kp.public().as_ref());
+                bytes.extend_from_slice(kp1.private().as_ref());
+            }
+            SuiKeyPair::Secp256k1KeyPair(kp) => {
+                let kp1 = kp.copy();
+                bytes.extend_from_slice(&[self.public().flag()]);
+                bytes.extend_from_slice(kp.public().as_ref());
+                bytes.extend_from_slice(kp1.private().as_ref());
+            }
+        }
+        base64ct::Base64::encode_string(&bytes[..])
+    }
+
+    fn decode_base64(value: &str) -> Result<Self, eyre::Report> {
+        let bytes =
+            base64ct::Base64::decode_vec(value).map_err(|e| eyre::eyre!("{}", e.to_string()))?;
+        match bytes.first() {
+            Some(x) => {
+                if x == &Ed25519SuiSignature::SCHEME.flag() {
+                    let sk = Ed25519PrivateKey::from_bytes(&bytes[1 + Ed25519PublicKey::LENGTH..])?;
+                    Ok(SuiKeyPair::Ed25519KeyPair(<Ed25519KeyPair as From<
+                        Ed25519PrivateKey,
+                    >>::from(sk)))
+                } else if x == &Secp256k1SuiSignature::SCHEME.flag() {
+                    let sk =
+                        Secp256k1PrivateKey::from_bytes(&bytes[1 + Secp256k1PublicKey::LENGTH..])?;
+                    Ok(SuiKeyPair::Secp256k1KeyPair(<Secp256k1KeyPair as From<
+                        Secp256k1PrivateKey,
+                    >>::from(
+                        sk
+                    )))
+                } else {
+                    Err(eyre::eyre!("Invalid flag byte"))
+                }
+            }
+            _ => Err(eyre::eyre!("Invalid bytes")),
+        }
+    }
+}
+
+impl AsRef<[u8]> for PublicKey<'_> {
+    fn as_ref(&self) -> &[u8] {
+        match self {
+            PublicKey::Ed25519KeyPair(pk) => pk.as_ref(),
+            PublicKey::Secp256k1KeyPair(pk) => pk.as_ref(),
+        }
+    }
+}
+
+impl PublicKey<'_> {
+    pub fn inner_into(&self) -> SuiAddress {
+        match *self {
+            PublicKey::Ed25519KeyPair(kp) => kp.into(),
+            PublicKey::Secp256k1KeyPair(kp) => kp.into(),
+        }
+    }
```

Comment:
Why is it not possible to implement `From<PublicKey<'_>> for SuiAddress`?

Reply:
related to the lifetime of PublicKey reference, gonna go with copy trait and simplify here